packages/wallet-sdk/src/derive.ts

```diff
@@ -71,19 +73,182 @@ export const deriveSalt = async (rootNode: BIP32Interface) => {
   return salt;
 };
 
+/**
+ * Determines how a private key is derived for an account:
+ *
+ * Wallet => STX_DERIVATION_PATH
+ * Data => DATA_DERIVATION_PATH
+ */
+export enum DerivationType {
+  Wallet,
+  Data,
+  Unknown,
+}
+
+/**
+ * Tries to find a derivation path for the stxPrivateKey for the account
+ * defined by rootNode and index that respects the username of that account.
+ *
+ * The stxPrivateKey is used to sign the profile of the account, therefore,
+ * a username must be owned by the stxPrivateKey.
+ *
+ * If a username is provided, a lookup for the owner address
+ * on the provided network is done.
+ *
+ * If no username is provided, a lookup for names owned
+ * by the stx derivation path and by the data derivation path is done.
+ * @param selectionOptions
+ * @returns username and derivation type
+ */
+export const selectStxDerivation = async ({
+  username,
+  rootNode,
+  index,
+  network,
+}: {
+  username?: string;
+  rootNode: BIP32Interface;
+  index: number;
+  network?: StacksNetwork;
+}): Promise<{ username: string | undefined; stxDerivationType: DerivationType }> => {
+  if (username) {
+    // Based on username, determine the derivation path for the stx private key
+    const stxDerivationTypeForUsername = await selectDerivationTypeForUsername({
+      username,
+      rootNode,
+      index,
+      network,
+    });
+    return { username, stxDerivationType: stxDerivationTypeForUsername };
+  } else {
+    const { username, derivationType } = await selectUsernameForAccount({
+      rootNode,
+      index,
+      network,
+    });
+    return { username, stxDerivationType: derivationType };
+  }
+};
+
+const selectDerivationTypeForUsername = async ({
+  username,
+  rootNode,
+  index,
+  network,
+}: {
+  username: string;
+  rootNode: BIP32Interface;
+  index: number;
+  network?: StacksNetwork;
+}): Promise<DerivationType> => {
+  if (network) {
+    const nameInfo = await network.getNameInfo(username);
+    let stxPrivateKey = deriveStxPrivateKey({ rootNode, index });
+    let derivedAddress = getAddressFromPrivateKey(stxPrivateKey);
+    if (derivedAddress !== nameInfo.address) {
+      // try data private key
+      stxPrivateKey = deriveDataPrivateKey({
+        rootNode,
+        index,
+      });
+      derivedAddress = getAddressFromPrivateKey(stxPrivateKey);
+      if (derivedAddress !== nameInfo.address) {
+        return DerivationType.Unknown;
+      } else {
+        return DerivationType.Data;
+      }
+    } else {
+      return DerivationType.Wallet;
+    }
+  } else {
+    // no network to determine the derivation path
+    return DerivationType.Unknown;
```

Comment:
Might it be better to just throw an error here?

Reply:
I agree an error is probably better than having a `DerivationType.Unknown`, especially because a lot of the uses of `DerivationType` just switch on whether it is equal to `Wallet`.

Reply:
I disagree, because we are processing a correctly created and registered name. The function should leave it to the user/developer of the library to handle this case. Maybe they have a fallback handling trying more derivation paths.